# Librerías

In [1]:
import os

# Librerias base para el analisis y manejo de datos
import numpy as np
import pandas as pd

pd.set_option('display.float_format', '{:.2f}'.format)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

# Librerias con los modelos principales
from sklearn.cluster import KMeans
from sklearn_extra.cluster import KMedoids 

# Libnreria para medir los resultados de los modelos
from sklearn.metrics import silhouette_score, adjusted_rand_score, calinski_harabasz_score

# Libreria para graficar
import matplotlib.pyplot as plt

# Estandarizar la data
from sklearn.preprocessing import StandardScaler

# Reducir la dimensionalidad para graficas
from sklearn.manifold import MDS

# metodo para recrear un iterador infinito
from itertools import cycle

from unidecode import unidecode

# Funciones

In [2]:
def convertir_col_numericas(df, lista_columnas):

    for columna in lista_columnas:

        df[columna] = df[columna].str.replace(',00', '', regex=False)
        df[columna] = df[columna].str.replace('.', '', regex=False)
        df[columna] = df[columna].str.replace(',', '.', regex=False).astype(float)

    return df

def porcentaje_participacion(df, lista_col, col_total):

    for col in lista_col:
        
        df[f'Porc_part_{col}'] = df[col] / df[col_total]

    return df


def crear_grafico_dispersion(modelo_fit, df, y):
    colors = plt.cm.tab20.colors + plt.cm.tab20b.colors
    color_cycle = cycle(colors)
    mds = MDS(n_components=2, dissimilarity = 'euclidean', random_state=123)
    df_transformado = mds.fit_transform(df)

    unique_labels = np.unique(modelo_fit.labels_)

    plt.figure(figsize=(10, 7))
    
    for idx, label in enumerate(unique_labels):
        color = next(color_cycle)  # Obtener el siguiente color del ciclo
        plt.scatter(df_transformado[modelo_fit.labels_ == label, 0], 
                    df_transformado[modelo_fit.labels_ == label, 1], 
                    label=f'Cluster {label}',
                    color=color)
    # Agregar las etiquetas de cada punto
    for i in range(df_transformado.shape[0]):
        plt.text(df_transformado[i, 0], 
                 df_transformado[i, 1], 
                 y.iloc[i, 0],  
                 fontsize=9, 
                 ha='right', 
                 color='black')

    plt.title('MDS Plot of Data with Labels')
    plt.xlabel('Dimension 1')
    plt.ylabel('Dimension 2')
    plt.legend()
    plt.show()

# Preparación bases de datos

## Base municipios

In [3]:
# Base Municipios
df_base = pd.read_excel("../Data/DIVIPOLA_Municipios.xlsx")

df_base['Departamento'] = df_base['Departamento'].apply(unidecode)
df_base['Municipio'] = df_base['Municipio'].apply(unidecode)
df_base['Municipio'] = df_base['Municipio'].replace('SAN ANDRES DE TUMACO', 'TUMACO')
df_base['Municipio'] = df_base['Municipio'].replace('SAN JOSE DE CUCUTA', 'CUCUTA')
df_base['Municipio'] = df_base['Municipio'].replace('CARTAGENA DE INDIAS', 'CARTAGENA')
df_base['Municipio'] = df_base['Municipio'].replace('VILLA DE SAN DIEGO DE UBATE', 'UBATE')
df_base['Municipio'] = df_base['Municipio'].replace('SANTIAGO DE CALI', 'CALI')
df_base['Municipio'] = df_base['Municipio'].replace('BOGOTA, D.C.', 'BOGOTA')
df_base['Departamento'] = df_base['Departamento'].replace('BOGOTA, D.C.', 'BOGOTA')

df_base['LLAVE'] = df_base['Departamento']+'_'+df_base['Municipio']

#
display(df_base.shape)
df_base.head(5)

(1122, 6)

CodigoD Departamento  CodigoM   Municipio       tipo                 LLAVE
0        5    ANTIOQUIA     5001    MEDELLIN  Municipio    ANTIOQUIA_MEDELLIN
1        5    ANTIOQUIA     5002   ABEJORRAL  Municipio   ANTIOQUIA_ABEJORRAL
2        5    ANTIOQUIA     5004    ABRIAQUI  Municipio    ANTIOQUIA_ABRIAQUI
3        5    ANTIOQUIA     5021  ALEJANDRIA  Municipio  ANTIOQUIA_ALEJANDRIA
4        5    ANTIOQUIA     5030       AMAGA  Municipio       ANTIOQUIA_AMAGA

## Base Inclusion financiera y saldos por municipio

Se agregan ambas bases en un solo dataset debido a que manejan los mismos indices para identificar municipios y departamentos

### Base inlcusion financiera

In [4]:
# Base Municipios
df_inclusion = pd.read_csv("../Data/Inclusi_n_Financiera_20250515.csv")

#
display(df_inclusion.shape)
df_inclusion.head(5)

(367775, 99)

TIPO_ENTIDAD  CODIGO_ENTIDAD          NOMBRE_ENTIDAD FECHA_CORTE  UNICAP  \
0             4              46         Coltefinanciera  31/03/2024      19   
1             1              59         Banco Santander  31/03/2024       4   
2             1              30  Banco Caja Social S.A.  31/03/2024      11   
3             1              13           BBVA Colombia  31/03/2024      27   
4             1              43              Banagrario  31/03/2024       2   

     DESCRIP_UC  RENGLON   DESC_RENGLON  \
0       QUINDIO      999  TOTAL QUINDÍO   
1       BOLIVAR      473        MORALES   
2  CUNDINAMARCA      126         CAJICA   
3      PUTUMAYO      760       SANTIAGO   
4     ATLANTICO      560       PONEDERA   

                                                TIPO  \
0                             Corresponsales fisicos   
1                             Corresponsales fisicos   
2                                  Cuentas de Ahorro   
3                             Corresponsales físicos   
4  Transacciones y tramites a traves de correspon...   

   (1) NRO_CORRESPONSALES_FÍSICOS_PROPIOS  \
0                                       1   
1                                       0   
2                                       0   
3                                       0   
4                                       0   

   (2) NRO_CORRESPONSALES_FÍSICOS_TERCERIZADOS  \
0                                            9   
1                                            7   
2                                            0   
3                                            0   
4                                            0   

   (3) NRO_CORRESPONSALES_FÍSICOS_ACTIVOS  (4) NRO_CORRESPONSALES_FÍSICOS  \
0                                      10                              10   
1                                       0                               7   
2                                       0                               0   
3                                       0                               0   
4                                       0                               0   

   (5) NRO_DEPÓSITOS_CORRESPONSALES_FÍSICOS  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                        42   

   (6) MONTO_DEPÓSITOS_CORRESPONSALES_FÍSICOS  \
0                                        0.00   
1                                        0.00   
2                                        0.00   
3                                        0.00   
4                                 62241730.00   

   (7) NRO_GIROS_ENVIADOS_CORRESPONSALES_FÍSICOS  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   

   (8) MONTO_GIROS_ENVIADOS_CORRESPONSALES_FÍSICOS  \
0                                             0.00   
1                                             0.00   
2                                             0.00   
3                                             0.00   
4                                             0.00   

   (9) NRO_GIROS_RECIBIDOS_CORRESPONSALES_FÍSICOS  \
0                                               0   
1                                               0   
2                                               0   
3                                               0   
4                                            2354   

   (10) MONTO_GIROS_RECIBIDOS_CORRESPONSALES_FÍSICOS  \
0                                               0.00   
1                                               0.00   
2                                               0.00   
3                                               0.00   
4                                      1248575811.00   

   (11) NRO_PAGOS_CORRESP

In [5]:
df_inclusion['TIPO'].unique()

array(['Corresponsales fisicos', 'Cuentas de Ahorro',
       'Corresponsales físicos',
       'Transacciones y tramites a traves de corresponsales fisicos',
       'Microcredito por rango', 'Credito de consumo',
       'Transacciones corresponsal movil', 'Microcredito',
       'Productos a nivel nacional', 'Credito de consumo de bajo monto',
       'Credito de vivienda', 'Numero de corresponsales moviles'],
      dtype=object)

In [6]:
# Se eliminan registros no objeto de estudio
col_eliminar = ['Corresponsales fisicos', 'Corresponsales físicos', 'Productos a nivel nacional', 'Microcredito por rango']

In [7]:
df_inclusion1 = df_inclusion[~df_inclusion['TIPO'].isin(col_eliminar)]
df_inclusion1 = df_inclusion1.drop(columns=[col for col in df_inclusion1.columns if 'HOMBRES' in col or 'MUJERES' in col or 'DEPÓSITOS' in col
                                           or 'GIROS' in col or 'PAGOS' in col or 'RETIROS' in col or 'TRANSFERENCIAS' in col or 'SMMLV' in col
                                           or 'ACTIVAS' in col or 'ACTIVOS' in col or 'PROPIOS' in col or 'TERCERIZADOS' in col])

In [8]:
df_inclusion1['Total_Numero'] = df_inclusion1[[col for col in df_inclusion1.columns if 'NRO' in col]].sum(axis=1)
df_inclusion1['Total_Monto'] = df_inclusion1[[col for col in df_inclusion1.columns if 'MONTO' in col]].sum(axis=1)
df_inclusion1['Total_Saldo'] = df_inclusion1[[col for col in df_inclusion1.columns if 'SALDO' in col]].sum(axis=1)

In [9]:
df_inclusion1.head(10)

TIPO_ENTIDAD  CODIGO_ENTIDAD             NOMBRE_ENTIDAD FECHA_CORTE  \
2              1              30     Banco Caja Social S.A.  31/03/2024   
4              1              43                 Banagrario  31/03/2024   
9              1              39           Banco Davivienda  31/03/2024   
12             1              57       Banco Pichincha S.A.  31/03/2024   
13             1              43                 Banagrario  31/03/2024   
17            32               5                    Confiar  31/03/2024   
18             1              39           Banco Davivienda  31/03/2024   
19             4             123     Bancar Tecnología C.F.  31/03/2024   
24             4             117               Credifamilia  31/03/2024   
25             1              42  Scotiabank Colpatria S.A.  31/03/2024   

    UNICAP       DESCRIP_UC  RENGLON          DESC_RENGLON  \
2       11     CUNDINAMARCA      126                CAJICA   
4        2        ATLANTICO      560              PONEDERA   
9        8            CAUCA      364               JAMBALO   
12       6           CALDAS      999          TOTAL CALDAS   
13      11     CUNDINAMARCA      372                 JUNIN   
17       5           BOYACA      798                 TENZA   
18      17           NARIÑO      215               CORDOBA   
19      24  VALLE DEL CAUCA      606              RESTREPO   
24       3           BOGOTA        1  SANTAFE DE BOGOTA D.   
25       8            CAUCA        1               POPAYAN   

                                                 TIPO  \
2                                   Cuentas de Ahorro   
4   Transacciones y tramites a traves de correspon...   
9   Transacciones y tramites a traves de correspon...   
12                                 Credito de consumo   
13                                  Cuentas de Ahorro   
17                   Transacciones corresponsal movil   
18  Transacciones y tramites a traves de correspon...   
19  Transacciones y tramites a traves de correspon...   
24  Transacciones y tramites a traves de correspon...   
25                                  Cuentas de Ahorro   

    (4) NRO_CORRESPONSALES_FÍSICOS  \
2                                0   
4                                0   
9                                0   
12                               0   
13                               0   
17                               0   
18                               0   
19                               0   
24                               0   
25                               0   

    (17) NRO_TRANSACCIONES_TRÁMITES_CORRESPONSALES_FÍSICOS  \
2                                                   0        
4                                                3142        
9                                                  57        
12                                                  0        
13                                                  0        
17                                                  0        
18                                               1052        
19                                                  2        
24                                              15474        
25                                                  0        

    (18) MONTO_TRANSACCIONES_CORRESPONSALES_FÍSICOS  (31) NRO_CTAS_AHORRO  \
2                                              0.00                 43198   
4                                     1971644602.00                     0   
9                                       13206705.00                     0   
12                                             0.00                     0   
13                                             0.00                  2976   
17                                             0.00                     0   
18                                     172835717.00                     0   
19                                        350000.00                     0   
24                                   10831417760.32                     

### Saldos por municipio

In [10]:
df_saldos = pd.read_csv("../Data/Saldo_de_las_captaciones_y_colocaciones_por_municipios_20250515.csv")

#
display(df_saldos.shape)
df_saldos.head(5)

(25146, 80)

Tipo de entidad  Código de entidad  Código del departamento  \
0                1                 54                        2   
1                2                 11                        2   
2                1                 65                       33   
3                4                 46                        2   
4                1                 56                        9   

   Código del municipio Nombre del municipio Fecha de Corte  \
0                   999      TOTAL ATLANTICO     31/12/2024   
1                   999      TOTAL ATLANTICO     31/12/2024   
2                   999        TOTAL VICHADA     31/12/2024   
3                   999      TOTAL ATLANTICO     31/12/2024   
4                     1           VALLEDUPAR     31/12/2024   

   Depósitos en cuenta corriente bancaria  Depósitos simples  \
0                          13720536515.29               0.00   
1                                    0.00               0.00   
2                                    0.00               0.00   
3                                    0.00               0.00   
4                              5213997.00               0.00   

   Certificados de depósito a término  Depósitos de ahorro  \
0                      95730055626.39       50706442658.93   
1                                0.00                 0.00   
2                                0.00          20634423.40   
3                      50466877248.13        1917691630.96   
4                      10033136304.00       10324833854.00   

   Cuenta de ahorros de valor real  Cuentas de ahorro especial  \
0                                0               1387137819.89   
1                                0                        0.00   
2                                0                        0.00   
3                                0                        0.00   
4                                0                        0.00   

   Certificado de ahorro valor real  Documentos por pagar  \
0                              0.00                  0.00   
1                              0.00                  0.00   
2                              0.00                  0.00   
3                              0.00                  0.00   
4                              0.00                  0.00   

   Cuenta centralizada  Títulos de inversión en circulación  \
0                    0                                 0.00   
1                    0                                 0.00   
2                    0                                 0.00   
3                    0                                 0.00   
4                    0                                 0.00   

   Cartera de créditos  Créditos de vivienda  Categoría A riesgo normal  \
0      369418264918.10        94732135062.00             87805425538.00   
1                 0.00                  0.00                       0.00   
2         105004921.06                  0.00                       0.00   
3        8884066294.77                  0.00                       0.00   
4       26485769389.00                  0.00                       0.00   

   Categoría B riesgo aceptable  Categoría C riesgo apreciable  \
0                 1621802675.00                  1215884648.00   
1                          0.00                           0.00   
2                          0.00                           0.00   
3                          0.00                           0.00   
4                          0.00                           0.00   

   Categoría D riesgo significativo  Categoría E riesgo de Incobrabilidad  \
0                     1736011736.00                         2353010465.00   
1                              0.00                                  0.00   
2                              0.00                                  0.00   
3                              0.00                                  0.00   
4                              0.00                                  0.00   

   Categoría A 

In [11]:
# Se eliminan variables no objeto de estudio
col_eliminar2 = ['Depósitos simples', 'Cuenta de ahorros de valor real', 'Certificado de ahorro valor real',
                 'Documentos por pagar', 'Cuenta centralizada', 'Títulos de inversión en circulación', 'Número de cuentas centralizadas',
                 'Número de oficinas', 'Número de empleados contratados', 'Número de empleados subcontratados', 'Componente contracíclico provisión individual',
                 'Derechos de transferencia de cartera de créditos por operaciones de apoyos transitorios de liquidez', 'Préstamos a empleados',
                 'Deterioro (provisión) préstamos a empleados', 'Número de CDT', 'Provisión créditos de vivienda', 'Provisión créditos y operaciones de leasing de consumo',
                 'Provisión microcréditos', 'Provisión créditos y operaciones de leasing comerciales']

In [12]:
df_saldos1 = df_saldos.drop(columns=col_eliminar2)

In [13]:
df_saldos1 = df_saldos1.drop(columns=[col for col in df_saldos1.columns if 'Categoría' in col])

In [14]:
df_saldos1.head(5)

Tipo de entidad  Código de entidad  Código del departamento  \
0                1                 54                        2   
1                2                 11                        2   
2                1                 65                       33   
3                4                 46                        2   
4                1                 56                        9   

   Código del municipio Nombre del municipio Fecha de Corte  \
0                   999      TOTAL ATLANTICO     31/12/2024   
1                   999      TOTAL ATLANTICO     31/12/2024   
2                   999        TOTAL VICHADA     31/12/2024   
3                   999      TOTAL ATLANTICO     31/12/2024   
4                     1           VALLEDUPAR     31/12/2024   

   Depósitos en cuenta corriente bancaria  Certificados de depósito a término  \
0                          13720536515.29                      95730055626.39   
1                                    0.00                                0.00   
2                                    0.00                                0.00   
3                                    0.00                      50466877248.13   
4                              5213997.00                      10033136304.00   

   Depósitos de ahorro  Cuentas de ahorro especial  Cartera de créditos  \
0       50706442658.93               1387137819.89      369418264918.10   
1                 0.00                        0.00                 0.00   
2          20634423.40                        0.00         105004921.06   
3        1917691630.96                        0.00        8884066294.77   
4       10324833854.00                        0.00       26485769389.00   

   Créditos de vivienda  Microcréditos, garantía idónea  \
0        94732135062.00                            0.00   
1                  0.00                            0.00   
2                  0.00                            0.00   
3                  0.00                            0.00   
4                  0.00                            0.00   

   Microcréditos, otras garantías  Provisión general  \
0                               0       947321350.62   
1                               0               0.00   
2                               0               0.00   
3                               0               0.00   
4                               0               0.00   

   Crédito de consumo de bajo monto  
0                              0.00  
1                              0.00  
2                              0.00  
3                              0.00  
4                              0.00

In [15]:
columnas_saldos = df_saldos1.columns[6:]

In [16]:
df_saldos2 = df_saldos1.groupby(['Código del departamento', 'Código del municipio'], as_index=False)[columnas_saldos].sum()

In [17]:
print(f"Cantidad de registros: {df_saldos2.shape[0]}")
df_saldos2.head(5)

Cantidad de registros: 1136


Código del departamento  Código del municipio  \
0                        1                     1   
1                        1                     2   
2                        1                     4   
3                        1                    21   
4                        1                    30   

   Depósitos en cuenta corriente bancaria  Certificados de depósito a término  \
0                       21518165177687.11                  256628043756287.44   
1                           5667716059.13                      87967288907.05   
2                                    0.00                                0.00   
3                          12752973094.99                       2581842338.00   
4                          46651221669.25                     110694770740.45   

   Depósitos de ahorro  Cuentas de ahorro especial  Cartera de créditos  \
0   129536376161561.78             827701702331.02   497346037428773.19   
1       89929330600.01                128179244.31      286420366741.25   
2           3750520.48                        0.00          93552322.25   
3        6627800442.44                        0.00       36445281479.87   
4      229420416256.72                690052690.12      379578586776.79   

   Créditos de vivienda  Microcréditos, garantía idónea  \
0     47272969911093.13                1680297796337.31   
1        37779672831.33                  71445711052.04   
2                  0.00                     87339818.88   
3          197401624.00                  30640797348.76   
4        73739136701.63                  36449474610.07   

   Microcréditos, otras garantías  Provisión general  \
0                               0   1536505567591.03   
1                               0       968353097.59   
2                               0               0.00   
3                               0       342945723.08   
4                               0       676195512.46   

   Crédito de consumo de bajo monto  
0                   418479804964.22  
1                              0.00  
2                              0.00  
3                              0.00  
4                              0.00

In [18]:
df_saldos2 = df_saldos2.rename(columns={
    'Código del departamento': 'UNICAP',
    'Código del municipio': 'RENGLON'
})

In [19]:
df_saldos2.head(1)

UNICAP  RENGLON  Depósitos en cuenta corriente bancaria  \
0       1        1                       21518165177687.11   

   Certificados de depósito a término  Depósitos de ahorro  \
0                  256628043756287.44   129536376161561.78   

   Cuentas de ahorro especial  Cartera de créditos  Créditos de vivienda  \
0             827701702331.02   497346037428773.19     47272969911093.13   

   Microcréditos, garantía idónea  Microcréditos, otras garantías  \
0                1680297796337.31                               0   

   Provisión general  Crédito de consumo de bajo monto  
0   1536505567591.03                   418479804964.22

### Union info financiera

In [20]:
# Paso 1: Crear dataframe base con valores únicos de DESCRIP_UC y DESC_RENGLON
df_inclusion2 = df_inclusion1[['DESCRIP_UC', 'DESC_RENGLON']].drop_duplicates().reset_index(drop=True)

# Paso 2: Iterar sobre cada valor único de TIPO
tipos_unicos = df_inclusion1['TIPO'].unique()

# Paso 3: Por cada TIPO, agrupar y renombrar columnas, luego hacer merge con df_base
for tipo in tipos_unicos:
    df_tmp = (
        df_inclusion1[df_inclusion1['TIPO'] == tipo]
        .groupby(['DESCRIP_UC','DESC_RENGLON'], as_index=False)[['Total_Numero', 'Total_Monto', 'Total_Saldo']]
        .sum()
    )
    
    # Paso 4: Renombrar columnas incluyendo el tipo
    df_tmp = df_tmp.rename(columns={
        'Total_Numero': f'Total_Numero_{tipo}',
        'Total_Monto': f'Total_Monto_{tipo}',
        'Total_Saldo': f'Total_Saldo_{tipo}'
    })
    
    # Paso 5: Merge al dataframe base
    df_inclusion2 = df_inclusion2.merge(df_tmp, on=['DESCRIP_UC', 'DESC_RENGLON'], how='left')



In [21]:
df_cod_inc = df_inclusion1[['UNICAP', 'DESCRIP_UC', 'RENGLON', 'DESC_RENGLON']].drop_duplicates().reset_index(drop=True)
df_cod_inc.shape

(1156, 4)

In [22]:
df_inc_cod = df_inclusion2.merge(df_cod_inc, on=['DESCRIP_UC', 'DESC_RENGLON'], how='left')

In [23]:
print(f"Cantidad de registros: {df_inc_cod.shape[0]}")
df_inc_cod.head(5)

Cantidad de registros: 1156


DESCRIP_UC  DESC_RENGLON  Total_Numero_Cuentas de Ahorro  \
0  CUNDINAMARCA        CAJICA                       658406.00   
1     ATLANTICO      PONEDERA                          489.00   
2         CAUCA       JAMBALO                        15396.00   
3        CALDAS  TOTAL CALDAS                      6342548.00   
4  CUNDINAMARCA         JUNIN                        12286.00   

   Total_Monto_Cuentas de Ahorro  Total_Saldo_Cuentas de Ahorro  \
0                           0.00               1411408475005.87   
1                           0.00                    58924254.26   
2                           0.00                 44340305658.85   
3                           0.00              10809986121402.48   
4                           0.00                 26766501997.40   

   Total_Numero_Transacciones y tramites a traves de corresponsales fisicos  \
0                                            1458152                          
1                                             120755                          
2                                              38996                          
3                                           16734970                          
4                                              68189                          

   Total_Monto_Transacciones y tramites a traves de corresponsales fisicos  \
0                                    512970951855.55                         
1                                     43006240891.49                         
2                                     17611943093.00                         
3                                   6770021531407.32                         
4                                     16444505526.00                         

   Total_Saldo_Transacciones y tramites a traves de corresponsales fisicos  \
0                                               0.00                         
1                                               0.00                         
2                                               0.00                         
3                                               0.00                         
4                                               0.00                         

   Total_Numero_Credito de consumo  Total_Monto_Credito de consumo  \
0                        227990.00                 190360743120.59   
1                           703.00                    161350593.30   
2                            98.00                    592142964.28   
3                       1438335.00                1788709787080.19   
4                           228.00                    651392100.65   

   Total_Saldo_Credito de consumo  \
0                            0.00   
1                            0.00   
2                            0.00   
3                            0.00   
4                            0.00   

   Total_Numero_Transacciones corresponsal movil  \
0                                         150.00   
1                                          56.00   
2                                           0.00   
3                                      118767.00   
4                                            NaN   

   Total_Monto_Transacciones corresponsal movil  \
0                                   31844847.00   
1                                    8821957.00   
2                                          0.00   
3                                31196446163.00   
4                                           NaN   

   Total_Saldo_Transacciones corresponsal movil  Total_Numero_Microcredito  \
0                                          0.00                     967.00   
1                                          0.00                       0.00   
2                                          0.00                     584.00   
3                                          0.00                   27664.00   
4                                           NaN                     360.00   

   Total_Monto_Microcredito  Total_Saldo_Microcredito  \
0       

In [24]:
# Se eliminan las siguientes columnas ya que dada la transformacion siempre van a ser NA o 0
col_no_validas = ['Total_Monto_Cuentas de Ahorro', 'Total_Saldo_Transacciones y tramites a traves de corresponsales fisicos',
                  'Total_Saldo_Credito de consumo', 'Total_Saldo_Transacciones corresponsal movil', 'Total_Saldo_Microcredito',
                  'Total_Saldo_Credito de consumo de bajo monto', 'Total_Saldo_Credito de vivienda', 'Total_Monto_Numero de corresponsales moviles',
                  'Total_Saldo_Numero de corresponsales moviles']

In [25]:
df_inclusion3 = df_inc_cod.drop(columns=col_no_validas)
df_inclusion3['DESCRIP_UC'] = df_inclusion3['DESCRIP_UC'].replace('NARIÑO', 'NARINO')
df_inclusion3['DESC_RENGLON'] = df_inclusion3['DESC_RENGLON'].replace('COVEÑAS', 'COVENAS')
df_inclusion3['DESC_RENGLON'] = df_inclusion3['DESC_RENGLON'].replace('(UBATE) VILLA DE SANDIEGO DE UBATE', 'UBATE')
df_inclusion3['DESC_RENGLON'] = df_inclusion3['DESC_RENGLON'].replace('DOS QUEBRADAS', 'DOSQUEBRADAS')
df_inclusion3['DESC_RENGLON'] = df_inclusion3['DESC_RENGLON'].replace('SANTAFE DE BOGOTA D.', 'BOGOTA')

df_inclusion3['LLAVE']=df_inclusion3['DESCRIP_UC']+'_'+df_inclusion3['DESC_RENGLON']

In [26]:
df_inclusion3['UNICAP'] = df_inclusion3['UNICAP'].astype(int)
df_inclusion3['RENGLON'] = df_inclusion3['RENGLON'].astype(int)

In [27]:
print(f"Cantidad de registros: {df_inclusion3.shape[0]}")
df_inclusion3.head(5)

Cantidad de registros: 1156


DESCRIP_UC  DESC_RENGLON  Total_Numero_Cuentas de Ahorro  \
0  CUNDINAMARCA        CAJICA                       658406.00   
1     ATLANTICO      PONEDERA                          489.00   
2         CAUCA       JAMBALO                        15396.00   
3        CALDAS  TOTAL CALDAS                      6342548.00   
4  CUNDINAMARCA         JUNIN                        12286.00   

   Total_Saldo_Cuentas de Ahorro  \
0               1411408475005.87   
1                    58924254.26   
2                 44340305658.85   
3              10809986121402.48   
4                 26766501997.40   

   Total_Numero_Transacciones y tramites a traves de corresponsales fisicos  \
0                                            1458152                          
1                                             120755                          
2                                              38996                          
3                                           16734970                          
4                                              68189                          

   Total_Monto_Transacciones y tramites a traves de corresponsales fisicos  \
0                                    512970951855.55                         
1                                     43006240891.49                         
2                                     17611943093.00                         
3                                   6770021531407.32                         
4                                     16444505526.00                         

   Total_Numero_Credito de consumo  Total_Monto_Credito de consumo  \
0                        227990.00                 190360743120.59   
1                           703.00                    161350593.30   
2                            98.00                    592142964.28   
3                       1438335.00                1788709787080.19   
4                           228.00                    651392100.65   

   Total_Numero_Transacciones corresponsal movil  \
0                                         150.00   
1                                          56.00   
2                                           0.00   
3                                      118767.00   
4                                            NaN   

   Total_Monto_Transacciones corresponsal movil  Total_Numero_Microcredito  \
0                                   31844847.00                     967.00   
1                                    8821957.00                       0.00   
2                                          0.00                     584.00   
3                                31196446163.00                   27664.00   
4                                           NaN                     360.00   

   Total_Monto_Microcredito  Total_Numero_Credito de consumo de bajo monto  \
0            14622740512.19                                           0.00   
1                      0.00                                           0.00   
2             4724967826.65                                           0.00   
3           222631986137.58                                           0.00   
4             3959467051.31                                           0.00   

   Total_Monto_Credito de consumo de bajo monto  \
0                                          0.00   
1                                          0.00   
2                                          0.00   
3                                          0.00   
4                                          0.00   

   Total_Numero_Credito de vivienda  Total_Monto_Credito de vivienda  \
0                            436.00                   57566457312.58   
1                              0.00                             0.00   
2                              0.00                             0.00   
3                           2759.00                  345598133641.79   
4                               NaN                              NaN   

   Total_Numero_Numero de corresponsales 

In [28]:
df_finan = df_inclusion3.merge(df_saldos2, on=['UNICAP', 'RENGLON'], how='left')

In [29]:
print(f"Cantidad de registros: {df_finan.shape[0]}")
df_finan.head(5)

Cantidad de registros: 1156


DESCRIP_UC  DESC_RENGLON  Total_Numero_Cuentas de Ahorro  \
0  CUNDINAMARCA        CAJICA                       658406.00   
1     ATLANTICO      PONEDERA                          489.00   
2         CAUCA       JAMBALO                        15396.00   
3        CALDAS  TOTAL CALDAS                      6342548.00   
4  CUNDINAMARCA         JUNIN                        12286.00   

   Total_Saldo_Cuentas de Ahorro  \
0               1411408475005.87   
1                    58924254.26   
2                 44340305658.85   
3              10809986121402.48   
4                 26766501997.40   

   Total_Numero_Transacciones y tramites a traves de corresponsales fisicos  \
0                                            1458152                          
1                                             120755                          
2                                              38996                          
3                                           16734970                          
4                                              68189                          

   Total_Monto_Transacciones y tramites a traves de corresponsales fisicos  \
0                                    512970951855.55                         
1                                     43006240891.49                         
2                                     17611943093.00                         
3                                   6770021531407.32                         
4                                     16444505526.00                         

   Total_Numero_Credito de consumo  Total_Monto_Credito de consumo  \
0                        227990.00                 190360743120.59   
1                           703.00                    161350593.30   
2                            98.00                    592142964.28   
3                       1438335.00                1788709787080.19   
4                           228.00                    651392100.65   

   Total_Numero_Transacciones corresponsal movil  \
0                                         150.00   
1                                          56.00   
2                                           0.00   
3                                      118767.00   
4                                            NaN   

   Total_Monto_Transacciones corresponsal movil  Total_Numero_Microcredito  \
0                                   31844847.00                     967.00   
1                                    8821957.00                       0.00   
2                                          0.00                     584.00   
3                                31196446163.00                   27664.00   
4                                           NaN                     360.00   

   Total_Monto_Microcredito  Total_Numero_Credito de consumo de bajo monto  \
0            14622740512.19                                           0.00   
1                      0.00                                           0.00   
2             4724967826.65                                           0.00   
3           222631986137.58                                           0.00   
4             3959467051.31                                           0.00   

   Total_Monto_Credito de consumo de bajo monto  \
0                                          0.00   
1                                          0.00   
2                                          0.00   
3                                          0.00   
4                                          0.00   

   Total_Numero_Credito de vivienda  Total_Monto_Credito de vivienda  \
0                            436.00                   57566457312.58   
1                              0.00                             0.00   
2                              0.00                             0.00   
3                           2759.00                  345598133641.79   
4                               NaN                              NaN   

   Total_Numero_Numero de corresponsales 

In [30]:
df_finan[df_finan['Depósitos en cuenta corriente bancaria'].isnull()].shape[1]

30

In [31]:
df_saldos = pd.read_csv("../Data/Saldo_de_las_captaciones_y_colocaciones_por_municipios_20250515.csv")

#
display(df_saldos.shape)
df_saldos.head(5)

(25146, 80)

Tipo de entidad  Código de entidad  Código del departamento  \
0                1                 54                        2   
1                2                 11                        2   
2                1                 65                       33   
3                4                 46                        2   
4                1                 56                        9   

   Código del municipio Nombre del municipio Fecha de Corte  \
0                   999      TOTAL ATLANTICO     31/12/2024   
1                   999      TOTAL ATLANTICO     31/12/2024   
2                   999        TOTAL VICHADA     31/12/2024   
3                   999      TOTAL ATLANTICO     31/12/2024   
4                     1           VALLEDUPAR     31/12/2024   

   Depósitos en cuenta corriente bancaria  Depósitos simples  \
0                          13720536515.29               0.00   
1                                    0.00               0.00   
2                                    0.00               0.00   
3                                    0.00               0.00   
4                              5213997.00               0.00   

   Certificados de depósito a término  Depósitos de ahorro  \
0                      95730055626.39       50706442658.93   
1                                0.00                 0.00   
2                                0.00          20634423.40   
3                      50466877248.13        1917691630.96   
4                      10033136304.00       10324833854.00   

   Cuenta de ahorros de valor real  Cuentas de ahorro especial  \
0                                0               1387137819.89   
1                                0                        0.00   
2                                0                        0.00   
3                                0                        0.00   
4                                0                        0.00   

   Certificado de ahorro valor real  Documentos por pagar  \
0                              0.00                  0.00   
1                              0.00                  0.00   
2                              0.00                  0.00   
3                              0.00                  0.00   
4                              0.00                  0.00   

   Cuenta centralizada  Títulos de inversión en circulación  \
0                    0                                 0.00   
1                    0                                 0.00   
2                    0                                 0.00   
3                    0                                 0.00   
4                    0                                 0.00   

   Cartera de créditos  Créditos de vivienda  Categoría A riesgo normal  \
0      369418264918.10        94732135062.00             87805425538.00   
1                 0.00                  0.00                       0.00   
2         105004921.06                  0.00                       0.00   
3        8884066294.77                  0.00                       0.00   
4       26485769389.00                  0.00                       0.00   

   Categoría B riesgo aceptable  Categoría C riesgo apreciable  \
0                 1621802675.00                  1215884648.00   
1                          0.00                           0.00   
2                          0.00                           0.00   
3                          0.00                           0.00   
4                          0.00                           0.00   

   Categoría D riesgo significativo  Categoría E riesgo de Incobrabilidad  \
0                     1736011736.00                         2353010465.00   
1                              0.00                                  0.00   
2                              0.00                                  0.00   
3                              0.00                                  0.00   
4                              0.00                                  0.00   

   Categoría A 

In [32]:
# Se eliminan variables no objeto de estudio
col_eliminar2 = ['Depósitos simples', 'Cuenta de ahorros de valor real', 'Certificado de ahorro valor real',
                 'Documentos por pagar', 'Cuenta centralizada', 'Títulos de inversión en circulación', 'Número de cuentas centralizadas',
                 'Número de oficinas', 'Número de empleados contratados', 'Número de empleados subcontratados', 'Componente contracíclico provisión individual',
                 'Derechos de transferencia de cartera de créditos por operaciones de apoyos transitorios de liquidez', 'Préstamos a empleados',
                 'Deterioro (provisión) préstamos a empleados', 'Número de CDT', 'Provisión créditos de vivienda', 'Provisión créditos y operaciones de leasing de consumo',
                 'Provisión microcréditos', 'Provisión créditos y operaciones de leasing comerciales']

In [33]:
df_saldos1 = df_saldos.drop(columns=col_eliminar2)

In [34]:
df_saldos1 = df_saldos1.drop(columns=[col for col in df_saldos1.columns if 'Categoría' in col])

In [35]:
df_saldos1.head(10)

Tipo de entidad  Código de entidad  Código del departamento  \
0                1                 54                        2   
1                2                 11                        2   
2                1                 65                       33   
3                4                 46                        2   
4                1                 56                        9   
5                1                 65                        5   
6                1                 39                        7   
7                1                 52                       20   
8                4                128                        5   
9                1                  1                       23   

   Código del municipio Nombre del municipio Fecha de Corte  \
0                   999      TOTAL ATLANTICO     31/12/2024   
1                   999      TOTAL ATLANTICO     31/12/2024   
2                   999        TOTAL VICHADA     31/12/2024   
3                   999      TOTAL ATLANTICO     31/12/2024   
4                     1           VALLEDUPAR     31/12/2024   
5                   491                NOBSA     31/12/2024   
6                     1            FLORENCIA     31/12/2024   
7                   682  SANTA ROSA DE CABAL     31/12/2024   
8                   212                COPER     31/12/2024   
9                     1               IBAGUE     31/12/2024   

   Depósitos en cuenta corriente bancaria  Certificados de depósito a término  \
0                          13720536515.29                      95730055626.39   
1                                    0.00                                0.00   
2                                    0.00                                0.00   
3                                    0.00                      50466877248.13   
4                              5213997.00                      10033136304.00   
5                                    0.00                                0.00   
6                          13816805318.40                      15605139719.09   
7                                    0.00                       4124095938.63   
8                                    0.00                                0.00   
9                          86808506895.86                     106880389631.00   

   Depósitos de ahorro  Cuentas de ahorro especial  Cartera de créditos  \
0       50706442658.93               1387137819.89      369418264918.10   
1                 0.00                        0.00                 0.00   
2          20634423.40                        0.00         105004921.06   
3        1917691630.96                        0.00        8884066294.77   
4       10324833854.00                        0.00       26485769389.00   
5          83674607.43                        0.00          53580892.81   
6      133791485504.89                  2229193.86      231318081213.47   
7        1086551854.12                        0.00        6092834208.72   
8           2136730.60                        0.00           4846045.18   
9      181175906229.01                310417661.99      689651543279.56   

   Créditos de vivienda  Microcréditos, garantía idónea  \
0        94732135062.00                            0.00   
1                  0.00                            0.00   
2                  0.00                            0.00   
3                  0.00                            0.00   
4                  0.00                            0.00   
5                  0.00                            0.00   
6        45523952850.20                            0.00   
7                  0.00                   5477250666.32   
8                  0.00                            0.00   
9       169494336850.29                            0.00   

   Microcréditos, otras garantías  Provisión general  \
0                               0       947321350.62   
1                               0               0.00   
2                               0             

## Mercado laboral

In [36]:
# Economia
df_mercadoLaboral = pd.read_csv("../Data/archivos_csv/Mercado_laboral.csv", sep=";", encoding="ISO-8859-1", low_memory=False)
df_mercadoLaboral = df_mercadoLaboral[df_mercadoLaboral['Año'] == 2016]
# 
display(df_mercadoLaboral.shape)
df_mercadoLaboral.head()

(25401, 13)

Código Departamento Departamento  Código Entidad         Entidad  \
7945                   99      Vichada           99624   Santa Rosalía   
7946                   99      Vichada           99773        Cumaribo   
7947                   99      Vichada           99524    La Primavera   
7948                   99      Vichada           99001  Puerto Carreño   
7949                   99      Vichada           99000         Vichada   

            Dimensión                                       Subcategoría  \
7945  Mercado laboral  Trabajadores cotizantes al sistema general de ...   
7946  Mercado laboral  Trabajadores cotizantes al sistema general de ...   
7947  Mercado laboral  Trabajadores cotizantes al sistema general de ...   
7948  Mercado laboral  Trabajadores cotizantes al sistema general de ...   
7949  Mercado laboral  Trabajadores cotizantes al sistema general de ...   

                                              Indicador Dato Numérico  \
7945  Trabajadores cotizantes al sistema general de ...        225,08   
7946  Trabajadores cotizantes al sistema general de ...      8.743,92   
7947  Trabajadores cotizantes al sistema general de ...        410,33   
7948  Trabajadores cotizantes al sistema general de ...      2.271,33   
7949  Trabajadores cotizantes al sistema general de ...     11.650,67   

      Dato Cualitativo   Año  Mes                         Fuente  \
7945               NaN  2016   12  FILCO- Ministerio del Trabajo   
7946               NaN  2016   12  FILCO- Ministerio del Trabajo   
7947               NaN  2016   12  FILCO- Ministerio del Trabajo   
7948               NaN  2016   12  FILCO- Ministerio del Trabajo   
7949               NaN  2016   12  FILCO- Ministerio del Trabajo   

     Unidad de Medida  
7945     Trabajadores  
7946     Trabajadores  
7947     Trabajadores  
7948     Trabajadores  
7949     Trabajadores

In [37]:
df_mercadoLaboral['Subcategoría'].unique()

array(['Trabajadores cotizantes al sistema general de seguridad social (total anual)',
       'Trabajadores cotizantes al sistema general de seguridad social  según grupos de edad',
       'Composición de los trabajadores cotizantes al sistema general de seguridad social  según grupos de edad',
       'Trabajadores cotizantes al sistema general de seguridad social  según sexo',
       'Composición de los trabajadores cotizantes al sistema general de seguridad social  según sexo',
       'Número de empresas generadoras de empleo formal por cada 10.000 habitantes',
       'Ocupación formal'], dtype=object)

In [38]:
lista_laboral = ['Porcentaje de personas ocupadas formalmente con respecto a la población total']

In [39]:
df_mercadoLaboral1 = df_mercadoLaboral[df_mercadoLaboral['Indicador'].isin(lista_laboral)]

In [40]:
# dejar datos por municipio / son 23 años
df_mercadoLaboral1 = df_mercadoLaboral1.pivot(
    index=['Código Entidad', 'Entidad'],
    columns='Indicador',
    values='Dato Numérico'
).reset_index()

In [41]:
df_mercadoLaboral1 = convertir_col_numericas(df_mercadoLaboral1, lista_laboral)

In [42]:
df_mercadoLaboral1 = df_mercadoLaboral1.rename(columns={
    'Porcentaje de personas ocupadas formalmente con respecto a la población total': 'Porc_trabajos_formales'
})

In [43]:
df_mercadoLaboral1['Porc_trabajos_formales'] = df_mercadoLaboral1['Porc_trabajos_formales'] / 100

In [44]:
df_mercadoLaboral1.head(5)

Indicador  Código Entidad     Entidad  Porc_trabajos_formales
0                    1001    Colombia                    0.28
1                    5001    Medellín                    0.62
2                    5002   Abejorral                    0.17
3                    5004    Abriaquí                    0.13
4                    5021  Alejandría                    0.14

## Transaccionalidad

In [45]:
df_transacciones = pd.read_csv("../Data/archivos_csv/transaccionalidad.csv", sep=",", encoding="utf-8")

In [46]:
display(df_transacciones.shape)
df_transacciones.head(3)

(5875015, 16)

TIPO_ENTIDAD  CODIGO_ENTIDAD    NOMBRE_ENTIDAD FECHA_CORTE  UNIDAD_CAPTURA  \
0             1              39  BANCO DAVIVIENDA  31/01/2023               2   
1             1              52     BANCAMÍA S.A.  31/01/2023               2   
2             1               7       BANCOLOMBIA  31/01/2023               2   

           NOMBRE_UNIDAD_CAPTURA  CODIGO_CANAL_DISTRIBUCION  \
0  Pagos por actividad económica                         14   
1  Pagos por actividad económica                         14   
2  Pagos por actividad económica                          9   

   CODIGO_SERRVICIO_TRANSACCIONAL  CODIGO_DEPARTAMENTO  CODIGO_MUNICIPIO  \
0                               0                   25             25175   
1                               0                   66             66170   
2                               0                   13             13006   

   CODIGO_OPERACION  CODIGO_TIPO_PERSONA  CODIGO_CIIU  \
0                 0                    0           17   
1                 0                    0          869   
2                 0                    0          475   

   CODIGO_MECANISMO_CONFIRMACION  CANTIDAD_OPERACIONES  MONTO_OPERACIONES  
0                              0                     2            1800000  
1                              0                     1             549000  
2                              0                     2            4230000

In [47]:
# Seleccionamos los codigos de operacion digitales
lista_operaciones_digitales = [1, 6, 7, 8, 13, 18]
df_transacciones = df_transacciones[df_transacciones['CODIGO_CANAL_DISTRIBUCION'].isin(lista_operaciones_digitales)]

In [48]:
df_transacciones1 = df_transacciones.groupby('CODIGO_MUNICIPIO', as_index=False)[['CANTIDAD_OPERACIONES', 'MONTO_OPERACIONES']].sum()


In [49]:
df_transacciones1.head(5)

CODIGO_MUNICIPIO  CANTIDAD_OPERACIONES  MONTO_OPERACIONES
0                 0           17661129015  13449395752819554
1              5001              35020083      2969120450068
2              5002                  3534          195297198
3              5004                    37            9128754
4              5021                   120            8469673

## Indice adopcion digital

### 2016

In [50]:
df_CNC_2016 = pd.read_csv("../Data/archivos_csv/d2016---copia.csv", sep=";", encoding="ISO-8859-1", low_memory=False)

df_CNC_2016.head(5)

ID              ESTUDIO  ENTREVISTADOR  REGISTRO    FECHAINI  \
0  2016  CC909403_AP_DIG_FIN           2368        18  15/07/2016   
1  2016  CC909403_AP_DIG_FIN           1826        19  15/07/2016   
2  2016  CC909403_AP_DIG_FIN           2380        24  15/07/2016   
3  2016  CC909403_AP_DIG_FIN           2380        32  15/07/2016   
4  2016  CC909403_AP_DIG_FIN           1826        33  15/07/2016   

         HORAINI    FECHAFIN        HORAFIN  DURACION TELEFONO   EMAIL  \
0  09:23:56 a.m.  15/07/2016  09:47:24 a.m.      1406   #NULL!  #NULL!   
1  09:38:23 a.m.  15/07/2016  10:12:49 a.m.      2064   #NULL!  #NULL!   
2  11:04:54 a.m.  15/07/2016  12:42:03 p.m.      5827   #NULL!  #NULL!   
3  04:10:30 p.m.  15/07/2016  06:00:53 p.m.      6622   #NULL!  #NULL!   
4  03:48:11 p.m.  15/07/2016  05:07:44 p.m.      4771   #NULL!  #NULL!   

   ID_CONTACTO  ESTADO  IDIOMA  PA  PB_MUN  PB_DANE  B1  PC  PD  KISH  \
0           -1       6       1   4      30    11001   1   2   5     4   
1           -1       6       1   4      30    11001   1   3   3     3   
2           -1       6       1   4      30    11001   1   2   5     2   
3           -1       6       1   4      30    11001   1   1   2     1   
4           -1       6       1   4      30    11001   1   1   3     3   

   PERS_NOMBRES_1  PERS_NOMBRES_2  PERS_NOMBRES_3  PERS_NOMBRES_4  \
0             NaN             NaN             NaN             NaN   
1             NaN             NaN             NaN             NaN   
2             NaN             NaN             NaN             NaN   
3             NaN             NaN             NaN             NaN   
4             NaN             NaN             NaN             NaN   

   PERS_NOMBRES_5  PERS_NOMBRES_6  PERS_NOMBRES_7  PERS_NOMBRES_8  \
0             NaN             NaN             NaN             NaN   
1             NaN             NaN             NaN             NaN   
2             NaN             NaN             NaN             NaN   
3             NaN             NaN             NaN             NaN   
4             NaN             NaN             NaN             NaN   

   PERS_SEXO_1 PERS_SEXO_2 PERS_SEXO_3 PERS_SEXO_4 PERS_SEXO_5 PERS_SEXO_6  \
0            1           2           1           2      #NULL!      #NULL!   
1            1           2           1      #NULL!      #NULL!      #NULL!   
2            2           1      #NULL!      #NULL!      #NULL!      #NULL!   
3            2      #NULL!      #NULL!      #NULL!      #NULL!      #NULL!   
4            1           1           1      #NULL!      #NULL!      #NULL!   

  PERS_SEXO_7 PERS_SEXO_8  PERS_EDAD_1 PERS_EDAD_2 PERS_EDAD_3 PERS_EDAD_4  \
0      #NULL!      #NULL!           34          35          40          60   
1      #NULL!      #NULL!           23          55          65      #NULL!   
2      #NULL!      #NULL!           30          45      #NULL!      #NULL!   
3      #NULL!      #NULL!           56      #NULL!      #NULL!      #NULL!   
4      #NULL!      #NULL!           22          38          57      #NULL!   

  PERS_EDAD_5 PERS_EDAD_6 PERS_EDAD_7 PERS_EDAD_8  DH             DIRECCION  \
0      #NULL!      #NULL!      #NULL!      #NULL!   5        CRA 97 56 G 15   
1      #NULL!      #NULL!      #NULL!      #NULL!   4     cra98  #56g14 sur   
2      #NULL!      #NULL!      #NULL!      #NULL!   6                #NULL!   
3      #NULL!      #NULL!      #NULL!      #NULL!   7     CLL 37 SUR #91_37   
4      #NULL!      #NULL!      #NULL!      #NULL!   5  CALLE 35SUR #90 A 15   

   RANGO_EDAD FECHA_ACTUAL       HORA_INI  PH  SECTOR  SECCION  MANZANA  \
0           6   15/07/2016  09:27:05 a.m.   2    4519        2       22   
1           6   15/07/2016  09:42:33 a.m.   2    4519        2       22   
2           5   15/07/2016  11:10:32 a.m.   1    4519        2       44   
3           6   15/07/2016  05:00:36 p.m.   1    4550        3        1   
4           2   15/07/2016  03:51:25 p.m.   1    4550        3       12   

   NOM_P1  S

In [51]:
col_2016 = ['PB_DANE', 'INDICADOR']
df_CNC_2016 = df_CNC_2016[col_2016]

In [52]:
df_CNC_2016['INDICADOR'] = df_CNC_2016['INDICADOR'].str.replace(',', '.', regex=False).astype(float)

In [53]:
df_CNC_2016.head(3)

PB_DANE  INDICADOR
0    11001       0.00
1    11001       0.00
2    11001       0.46

In [54]:
df_CNC_2016_1 = df_CNC_2016.groupby('PB_DANE', as_index=False)[['INDICADOR']].mean()
df_CNC_2016_1['Año_datos'] = 2016

In [55]:
df_CNC_2016_1.head(3)

PB_DANE  INDICADOR  Año_datos
0     5001       0.14       2016
1     5031       0.01       2016
2     5045       0.08       2016

### 2018

In [56]:
df_CNC_2018 = pd.read_csv("../Data/archivos_csv/d2018---copia.csv", sep=";", encoding="ISO-8859-1", low_memory=False)

df_CNC_2018.head(5)

ID               ESTUDIO  ENTREVISTADOR  REGISTRO    FECHAINI  \
0  2018  CC910501_2017_APD_II           5338         5  16/11/2017   
1  2018  CC910501_2017_APD_II           3804         8  16/11/2017   
2  2018  CC910501_2017_APD_II           3804         9  16/11/2017   
3  2018  CC910501_2017_APD_II           3804        10  16/11/2017   
4  2018  CC910501_2017_APD_II           2125        12  16/11/2017   

         HORAINI    FECHAFIN        HORAFIN  DURACION  TELEFONO  EMAIL  \
0  10:40:05 a.m.  16/11/2017  11:20:35 a.m.      2429       NaN    NaN   
1  01:34:56 p.m.  16/11/2017  02:06:50 p.m.      1913       NaN    NaN   
2  02:17:33 p.m.  16/11/2017  02:38:54 p.m.      1280       NaN    NaN   
3  03:43:34 p.m.  16/11/2017  04:19:27 p.m.      2153       NaN    NaN   
4  04:45:42 p.m.  16/11/2017  05:44:01 p.m.      3497       NaN    NaN   

   ID_CONTACTO  ESTADO  IDIOMA  GPSVAL  GPSALT  GPSLONG  GPSLAT  GPSPRECISION  \
0           -1       6  #NULO!     NaN     NaN      NaN     NaN           NaN   
1           -1       6  #NULO!     NaN     NaN      NaN     NaN           NaN   
2           -1       6  #NULO!     NaN     NaN      NaN     NaN           NaN   
3           -1       6  #NULO!     NaN     NaN      NaN     NaN           NaN   
4           -1       6  #NULO!     NaN     NaN      NaN     NaN           NaN   

  IDAPK IDBROWSER IDDEVICE IDMODE IDMODEL IDVERSO  IDVERSION IDUSUARIO  \
0   NaN       NaN      NaN    NaN     NaN     NaN         15       NaN   
1   NaN       NaN      NaN    NaN     NaN     NaN         16       NaN   
2   NaN       NaN      NaN    NaN     NaN     NaN         16       NaN   
3   NaN       NaN      NaN    NaN     NaN     NaN         16       NaN   
4   NaN       NaN      NaN    NaN     NaN     NaN         17       NaN   

    IDEND        DIREC_VIVI  DH  EDAD  RANGO_EDAD  PA     PB  PB1  N_ENCUESTA  \
0  #NULO!    Kr 14 # 21 105   5    69           7   2   8001    1      717377   
1  #NULO!  kra 46A # 110_64   4    56           6   1   5001    1      480118   
2  #NULO!  Kra 46A # 110_60   0    64           6   1   5001    1      505241   
3  #NULO!   kra 109 201 _28   8    38           4   1   5001    1      946330   
4  #NULO!    Cll 58 47 aw 8   8    43           4   5  68001    1      416040   

   LAT_1  LONG_1  PC  PD  KISH FECHA_ACTUAL       HORA_INI  PH  SECTOR  \
0  10.95  -74.78   1   1     1   16/11/2017  10:41:54 a.m.   2    2201   
1    NaN     NaN   1   5     4   16/11/2017  01:37:53 p.m.   2     203   
2    NaN     NaN   1   4     4   16/11/2017  02:20:00 p.m.   2     203   
3   6.30  -75.55   2   2     2   16/11/2017  03:45:34 p.m.   2     203   
4   7.09  -73.15   1   4     3   16/11/2017  04:47:44 p.m.   2     798   

   SECCION  MANZANA  XPD_1  P1_NOMBRES_1  P1_NOMBRES_2  P1_NOMBRES_3  \
0        8        3    NaN           NaN           NaN           NaN   
1        1       13    NaN           NaN           NaN           NaN   
2        1       13    NaN           NaN           NaN           NaN   
3        2        2    NaN           NaN           NaN           NaN   
4        2       25    NaN           NaN           NaN           NaN   

   P1_NOMBRES_4  P1_NOMBRES_5  P1_NOMBRES_6  P1_NOMBRES_7  P1_NOMBRES_8  \
0           NaN           NaN           NaN           NaN           NaN   
1           NaN           NaN           NaN           NaN           NaN   
2           NaN           NaN           NaN           NaN           NaN   
3           NaN           NaN           NaN           NaN           NaN   
4           NaN           NaN           NaN           NaN           NaN   

   P1_NOMBRES_9  P1_NOMBRES_10  P1_NOMBRES_11  P1_NOMBRES_12  P1_SEXO_1  \
0           NaN            NaN            NaN            NaN          2   
1           NaN            NaN            NaN            NaN          2   
2           NaN            NaN            NaN            NaN          2   
3           NaN            NaN            NaN            NaN          1   
4           NaN 

In [57]:
col_2018 = ['PB', 'INDICADOR']
df_CNC_2018 = df_CNC_2018[col_2018]

In [58]:
df_CNC_2018_1 = df_CNC_2018.groupby('PB', as_index=False)[['INDICADOR']].mean()
df_CNC_2018_1['Año_datos'] = 2018

In [59]:
df_CNC_2018_1 = df_CNC_2018_1.rename(columns={
    'PB': 'PB_DANE'
})

In [60]:
df_CNC_2016_1.head(3)

PB_DANE  INDICADOR  Año_datos
0     5001       0.14       2016
1     5031       0.01       2016
2     5045       0.08       2016

### 2023

In [61]:
df_CNC_2023 = pd.read_csv("../Data/archivos_csv/d2023---copia.csv", sep=";", encoding="ISO-8859-1", low_memory=False)

df_CNC_2023.head(5)

ID               ESTUDIO  ENTREVISTADOR  REGISTRO   FECHAINI   HORAINI  \
0  2023  CC913901_AD_23_HOGAR          27848      1877  4/04/2023  14:47:01   
1  2023  CC913901_AD_23_HOGAR          16506      1885  4/04/2023  15:10:56   
2  2023  CC913901_AD_23_HOGAR           8681      1889  4/04/2023  15:16:08   
3  2023  CC913901_AD_23_HOGAR           8910      1898  4/04/2023  15:54:46   
4  2023  CC913901_AD_23_HOGAR           5739      1901  4/04/2023  15:59:03   

    FECHAFIN   HORAFIN  DURACION  TELEFONO  EMAIL  ID_CONTACTO  ESTADO  \
0  4/04/2023  16:14:40      5257       NaN    NaN            1       6   
1  4/04/2023  16:31:17      4821       NaN    NaN            1       6   
2  4/04/2023  16:02:38      2758       NaN    NaN            1       6   
3  4/04/2023  17:03:58      4150       NaN    NaN            1       6   
4  4/04/2023  17:14:08      4504       NaN    NaN            1       6   

   IDIOMA     IDCOORD  GPSVAL                  GPSALT      GPSLONG  \
0       1  ggutierrez       1   9.219.000.244.140.620   73,1366221   
1       1  ggutierrez       1                     322   72,5136372   
2       1  ggutierrez       1  32.079.998.779.296.800    72,513735   
3       1  ggutierrez       0                 211.498  72,53553608   
4       1  ggutierrez       1   2.983.999.938.964.840   72,5361709   

        GPSLAT  GPSPRECISION                            IDGUID  \
0    7,1137993         14.00  68A6DDD4423C4F71BA811C93CA1D061C   
1    7,8924644         11.00  B466F09F0CE8409987CA519E12CF3DE0   
2    7,8923787         10.00  FD9C044BC5BA4EF1B0F64E4D2E0CE7A0   
3  7,899647817          5.00  FD03B7DD2C2A4D68A41F4F6C6C64851A   
4    7,8981501          4.00  6FA11A5A3480474BB119185CE8D534A0   

   IDGUIDCONTACTO                         IDGUIDREC       IDAPK IDBROWSER  \
0             NaN  F528C6073E4C44A9A0DE4B8935316B41  4.4.2142.1    Chrome   
1             NaN  CD3AB28CEEEA45BB89047C99D8F2F3A4  4.4.2142.1    Chrome   
2             NaN  83875966BBF1442183C5CD744FC2C42F  4.4.2142.1    Chrome   
3             NaN  FEF6A873B7C540B1AA139756F2B3E238  4.4.2130.3    Chrome   
4             NaN  9C54F344C3044860B22BB127AE2B2E92  4.4.2142.1    Chrome   

           IDDEVICE  IDIP IDMODE  IDMODEL    IDSCREEN     IDSO  \
0  6ad3714b108c5ad7   NaN      E  SMT285M   601;962;V  Android   
1  5b68ffd465f9e035   NaN      E   SMT225  601;1007;V  Android   
2  0e4bc59f3d2ac538   NaN      E   SMT225   894;534;H  Android   
3  287d5de6ba10d0e4   NaN      E   BG2U03   451;770;V  Android   
4  f72b0d598143cba8   NaN      E   SMT225  601;1007;V  Android   

                    IDPPI  IDVERBROWSER IDVERSO  IDVERSION   IDUSUARIO  \
0  12.780.000.686.645.500            95   5.1.1         16    63544993   
1  12.780.000.686.645.500           108      11         16  1093785708   
2                     144           108      12         16  1093793845   
3  12.780.000.686.645.500            64       7         16  1093734724   
4  12.780.000.686.645.500           108      12         16    60450198   

   TIPO_ENCUESTA  TIPO  AUX99_1  IDEND  N_ENCUESTA  VERSION  GPSE  \
0              1     3        1      1      764678       16     1   
1              1     3        1      1      900989       16     1   
2              1     3        1      1      968646       16     1   
3              1     3        1      1      180785       16     0   
4              1     3        1      1      734690       16     1   

        LONG_1        LAT_1  VERIFICA FOTO  REGIONAL  PB1  ID_MANZANA  \
0   73,1366221    7,1137993         1  NaN         5    1      308055   
1   72,5136372    7,8924644         1  NaN         5    1      282339   
2    72,513735    7,8923787         1  NaN         5    1      282339   
3  72,53553608  7,899647817         1  NaN         5    1      282690   
4   72,5361709    7,8981501         1  NaN         5    1      282690   

   ID_MANZ_AUX  AUX99      CODDANE      POINT_X      POINT_Y  BAN  SECTOR  \
0       308055      1  6,80011E

In [62]:
col_2023 = ['DANE', 'indicador']
df_CNC_2023 = df_CNC_2023[col_2023]

In [63]:
df_CNC_2023['indicador'] = df_CNC_2023['indicador'].str.strip()
df_CNC_2023['indicador'] = df_CNC_2023['indicador'].str.replace('-', '0', regex=False)
df_CNC_2023['indicador'] = df_CNC_2023['indicador'].str.replace(',', '.', regex=False).astype(float)

In [64]:
df_CNC_2023_1 = df_CNC_2023.groupby('DANE', as_index=False)[['indicador']].mean()
df_CNC_2023_1['Año_datos'] = 2023

In [65]:
df_CNC_2023_1 = df_CNC_2023_1.rename(columns={
    'DANE': 'PB_DANE',
    'indicador': 'INDICADOR'
})

In [66]:
df_CNC_2023_1.head(3)

PB_DANE  INDICADOR  Año_datos
0     5001       0.31       2023
1     5031       0.29       2023
2     5045       0.17       2023

### Validacion columnas

In [67]:
print(df_CNC_2016_1.columns, df_CNC_2016_1.shape[0])
print(df_CNC_2018_1.columns, df_CNC_2018_1.shape[0])
print(df_CNC_2023_1.columns, df_CNC_2023_1.shape[0])

Index(['PB_DANE', 'INDICADOR', 'Año_datos'], dtype='object') 61
Index(['PB_DANE', 'INDICADOR', 'Año_datos'], dtype='object') 46
Index(['PB_DANE', 'INDICADOR', 'Año_datos'], dtype='object') 55


### Union todos los datasets

In [68]:
cnc_concat = pd.concat([df_CNC_2016_1, df_CNC_2018_1, df_CNC_2023_1], ignore_index=True)

In [69]:
cnc_concat.shape[0]

162

### Seleccion valor mas actualizado por municipio

In [70]:
# Ordena por cod_municipio y año (de más reciente a más antiguo)
cnc_concat1 = cnc_concat.sort_values(by=['PB_DANE', 'Año_datos'], ascending=[True, False])

# Elimina duplicados, conservando solo el más reciente por cod_municipio
cnc_concat1 = cnc_concat1.drop_duplicates(subset='PB_DANE', keep='first')

In [71]:
display(cnc_concat1.shape)
cnc_concat1.head(3)

(82, 3)

PB_DANE  INDICADOR  Año_datos
107     5001       0.31       2023
108     5031       0.29       2023
109     5045       0.17       2023

In [72]:
cnc_concat_def = cnc_concat1[['PB_DANE', 'INDICADOR']]

# Armado dataframe Final

## Agregar info depto

In [73]:
df1 = pd.merge(cnc_concat_def, df_base[['CodigoD','CodigoM', 'Municipio','Departamento','LLAVE']], left_on='PB_DANE', right_on='CodigoM', how='left')

In [74]:
df1 = df1.drop(['CodigoM'], axis=1)

In [75]:
display(df1.shape)
df1.head(3)

(82, 6)

PB_DANE  INDICADOR  CodigoD Municipio Departamento               LLAVE
0     5001       0.31        5  MEDELLIN    ANTIOQUIA  ANTIOQUIA_MEDELLIN
1     5031       0.29        5    AMALFI    ANTIOQUIA    ANTIOQUIA_AMALFI
2     5045       0.17        5  APARTADO    ANTIOQUIA  ANTIOQUIA_APARTADO

## Info Mercado Laboral

In [76]:
df6 = pd.merge(df1, df_mercadoLaboral1[['Código Entidad','Porc_trabajos_formales']], left_on='PB_DANE', right_on='Código Entidad', how='left')

In [77]:
df6 = df6.drop(['Código Entidad'], axis=1)
#df6 = df6.fillna(0)

In [78]:
display(df6.shape)
df6.head(3)

(82, 7)

PB_DANE  INDICADOR  CodigoD Municipio Departamento               LLAVE  \
0     5001       0.31        5  MEDELLIN    ANTIOQUIA  ANTIOQUIA_MEDELLIN   
1     5031       0.29        5    AMALFI    ANTIOQUIA    ANTIOQUIA_AMALFI   
2     5045       0.17        5  APARTADO    ANTIOQUIA  ANTIOQUIA_APARTADO   

   Porc_trabajos_formales  
0                    0.62  
1                    0.12  
2                    0.29

## Info Transaccionalidad

In [79]:
df7 = pd.merge(df6, df_transacciones1, left_on='PB_DANE', right_on='CODIGO_MUNICIPIO', how='left')

In [80]:
df7 = df7.drop(['CODIGO_MUNICIPIO'], axis=1)

In [81]:
display(df7.shape)
df7.head(3)

(82, 9)

PB_DANE  INDICADOR  CodigoD Municipio Departamento               LLAVE  \
0     5001       0.31        5  MEDELLIN    ANTIOQUIA  ANTIOQUIA_MEDELLIN   
1     5031       0.29        5    AMALFI    ANTIOQUIA    ANTIOQUIA_AMALFI   
2     5045       0.17        5  APARTADO    ANTIOQUIA  ANTIOQUIA_APARTADO   

   Porc_trabajos_formales  CANTIDAD_OPERACIONES  MONTO_OPERACIONES  
0                    0.62              35020083      2969120450068  
1                    0.12                  1836          128857908  
2                    0.29                  5145         1324013986

## Info financiera

In [82]:
df8 = pd.merge(df7, df_finan, left_on='LLAVE', right_on='LLAVE', how='left')

In [83]:
display(df8.shape)
df8.head(3)

(82, 38)

PB_DANE  INDICADOR  CodigoD Municipio Departamento               LLAVE  \
0     5001       0.31        5  MEDELLIN    ANTIOQUIA  ANTIOQUIA_MEDELLIN   
1     5031       0.29        5    AMALFI    ANTIOQUIA    ANTIOQUIA_AMALFI   
2     5045       0.17        5  APARTADO    ANTIOQUIA  ANTIOQUIA_APARTADO   

   Porc_trabajos_formales  CANTIDAD_OPERACIONES  MONTO_OPERACIONES DESCRIP_UC  \
0                    0.62              35020083      2969120450068  ANTIOQUIA   
1                    0.12                  1836          128857908  ANTIOQUIA   
2                    0.29                  5145         1324013986  ANTIOQUIA   

  DESC_RENGLON  Total_Numero_Cuentas de Ahorro  Total_Saldo_Cuentas de Ahorro  \
0     MEDELLIN                     25943055.00             129671242139741.66   
1       AMALFI                        35986.00                 60798059764.96   
2     APARTADO                      1257644.00               1133602316218.53   

   Total_Numero_Transacciones y tramites a traves de corresponsales fisicos  \
0                                           72079011                          
1                                             446850                          
2                                            3293454                          

   Total_Monto_Transacciones y tramites a traves de corresponsales fisicos  \
0                                  27413681315996.00                         
1                                    243367748265.00                         
2                                   1690522134159.52                         

   Total_Numero_Credito de consumo  Total_Monto_Credito de consumo  \
0                       7340898.00               14623377570065.94   
1                         11528.00                   9556225658.60   
2                        103939.00                 248789293580.21   

   Total_Numero_Transacciones corresponsal movil  \
0                                       33688.00   
1                                           0.00   
2                                          27.00   

   Total_Monto_Transacciones corresponsal movil  Total_Numero_Microcredito  \
0                                 7009122029.00                   29907.00   
1                                          0.00                     274.00   
2                                    3926524.00                    2853.00   

   Total_Monto_Microcredito  Total_Numero_Credito de consumo de bajo monto  \
0           342432552857.99                                      288743.00   
1             3031182200.00                                           0.00   
2            29134726425.21                                           0.00   

   Total_Monto_Credito de consumo de bajo monto  \
0                               225005586131.00   
1                                          0.00   
2                                          0.00   

   Total_Numero_Credito de vivienda  Total_Monto_Credito de vivienda  \
0                          24126.00                 3833395587878.78   
1                             30.00                    4211532354.00   
2                            462.00                   31254863802.85   

   Total_Numero_Numero de corresponsales moviles  UNICAP  RENGLON  \
0                                         424.00       1        1   
1                                           4.00       1       31   
2                                          15.00       1       45   

   Depósitos en cuenta corriente bancaria  Certificados de depósito a término  \
0                       21518165177687.11                  256628043756287.44   
1                          30697763679.92                      49900031440.51   
2                         300272905199.44                     280486764292.02   

   Depósitos de ahorro  Cuentas de ahorro especial  Cartera de créditos  \
0   129536376161561.78             827701702331.02   497346037428773.19   
1       60841194925.44                122287054.12  

In [84]:
df9 = df8.drop(columns=['CodigoD', 'Municipio', 'Departamento', 'LLAVE', 'UNICAP','RENGLON', 'DESCRIP_UC', 'DESC_RENGLON'])

nulos = df9['Depósitos en cuenta corriente bancaria'].isnull().sum()

print(f'Dimensiones finales del dataset: {df9.shape}')
print(f'Cantidad de registros nulos: {nulos}')
df9[df9['Depósitos en cuenta corriente bancaria'].isnull()]

Dimensiones finales del dataset: (82, 30)
Cantidad de registros nulos: 0


Empty DataFrame
Columns: [PB_DANE, INDICADOR, Porc_trabajos_formales, CANTIDAD_OPERACIONES, MONTO_OPERACIONES, Total_Numero_Cuentas de Ahorro, Total_Saldo_Cuentas de Ahorro, Total_Numero_Transacciones y tramites a traves de corresponsales fisicos, Total_Monto_Transacciones y tramites a traves de corresponsales fisicos, Total_Numero_Credito de consumo, Total_Monto_Credito de consumo, Total_Numero_Transacciones corresponsal movil, Total_Monto_Transacciones corresponsal movil, Total_Numero_Microcredito, Total_Monto_Microcredito, Total_Numero_Credito de consumo de bajo monto, Total_Monto_Credito de consumo de bajo monto, Total_Numero_Credito de vivienda, Total_Monto_Credito de vivienda, Total_Numero_Numero de corresponsales moviles, Depósitos en cuenta corriente bancaria, Certificados de depósito a término, Depósitos de ahorro, Cuentas de ahorro especial, Cartera de créditos, Créditos de vivienda, Microcréditos, garantía idónea, Microcréditos, otras garantías, Provisión general, Crédito de consumo de bajo monto]
Index: []

In [85]:
df9.head(5)

PB_DANE  INDICADOR  Porc_trabajos_formales  CANTIDAD_OPERACIONES  \
0     5001       0.31                    0.62              35020083   
1     5031       0.29                    0.12                  1836   
2     5045       0.17                    0.29                  5145   
3     5147       0.22                    0.11                   665   
4     5266       0.44                    0.26               1535208   

   MONTO_OPERACIONES  Total_Numero_Cuentas de Ahorro  \
0      2969120450068                     25943055.00   
1          128857908                        35986.00   
2         1324013986                      1257644.00   
3           45517947                       144053.00   
4       936636951771                      1748964.00   

   Total_Saldo_Cuentas de Ahorro  \
0             129671242139741.66   
1                 60798059764.96   
2               1133602316218.53   
3                144877378126.73   
4               4751826417377.74   

   Total_Numero_Transacciones y tramites a traves de corresponsales fisicos  \
0                                           72079011                          
1                                             446850                          
2                                            3293454                          
3                                            1358494                          
4                                            4443206                          

   Total_Monto_Transacciones y tramites a traves de corresponsales fisicos  \
0                                  27413681315996.00                         
1                                    243367748265.00                         
2                                   1690522134159.52                         
3                                    650252705224.00                         
4                                   2173654183583.00                         

   Total_Numero_Credito de consumo  Total_Monto_Credito de consumo  \
0                       7340898.00               14623377570065.94   
1                         11528.00                   9556225658.60   
2                        103939.00                 248789293580.21   
3                         13105.00                  22873702668.14   
4                       1094451.00                1701603505519.54   

   Total_Numero_Transacciones corresponsal movil  \
0                                       33688.00   
1                                           0.00   
2                                          27.00   
3                                          33.00   
4                                          17.00   

   Total_Monto_Transacciones corresponsal movil  Total_Numero_Microcredito  \
0                                 7009122029.00                   29907.00   
1                                          0.00                     274.00   
2                                    3926524.00                    2853.00   
3                                    5879874.00                    1351.00   
4                                     815544.00                    2022.00   

   Total_Monto_Microcredito  Total_Numero_Credito de consumo de bajo monto  \
0           342432552857.99                                      288743.00   
1             3031182200.00                                           0.00   
2            29134726425.21                                           0.00   
3            10511834181.94                                           0.00   
4            24514106163.54                                           0.00   

   Total_Monto_Credito de consumo de bajo monto  \
0                               225005586131.00   
1                                          0.00   
2                                          0.00   
3                                          0.00   
4                                          0.00   

   Total_Numero_Credito de vivienda  Total_Monto_Credito de vivienda  \
0                          

## Estructura final entrenamiento modelo

In [87]:
# Columnas a escalar
col_estandarizar = df9.columns[3:]

# Inicializar el escalador
scaler = StandardScaler()

# Aplicar el escalado solo a las columnas seleccionadas
df_scaled_part = pd.DataFrame(
    scaler.fit_transform(df9[col_estandarizar]),
    columns=col_estandarizar,
    index=df9.index
)

# Unir las columnas no escaladas con las escaladas
df_scaled = pd.concat([df9[df9.columns[:3]], df_scaled_part], axis=1)

# Mostrar resultado
df_scaled.head(5)

PB_DANE  INDICADOR  Porc_trabajos_formales  CANTIDAD_OPERACIONES  \
0     5001       0.31                    0.62                  0.84   
1     5031       0.29                    0.12                 -0.13   
2     5045       0.17                    0.29                 -0.13   
3     5147       0.22                    0.11                 -0.13   
4     5266       0.44                    0.26                 -0.09   

   MONTO_OPERACIONES  Total_Numero_Cuentas de Ahorro  \
0               0.17                            1.97   
1              -0.13                           -0.27   
2              -0.13                           -0.16   
3              -0.13                           -0.26   
4              -0.04                           -0.12   

   Total_Saldo_Cuentas de Ahorro  \
0                           1.66   
1                          -0.18   
2                          -0.17   
3                          -0.18   
4                          -0.11   

   Total_Numero_Transacciones y tramites a traves de corresponsales fisicos  \
0                                               3.42                          
1                                              -0.31                          
2                                              -0.16                          
3                                              -0.26                          
4                                              -0.10                          

   Total_Monto_Transacciones y tramites a traves de corresponsales fisicos  \
0                                               4.45                         
1                                              -0.35                         
2                                              -0.10                         
3                                              -0.28                         
4                                              -0.01                         

   Total_Numero_Credito de consumo  Total_Monto_Credito de consumo  \
0                             0.35                            1.47   
1                            -0.15                           -0.19   
2                            -0.14                           -0.17   
3                            -0.15                           -0.19   
4                            -0.07                           -0.00   

   Total_Numero_Transacciones corresponsal movil  \
0                                           0.24   
1                                          -0.26   
2                                          -0.26   
3                                          -0.26   
4                                          -0.26   

   Total_Monto_Transacciones corresponsal movil  Total_Numero_Microcredito  \
0                                          0.26                       0.95   
1                                         -0.28                      -0.39   
2                                         -0.28                      -0.27   
3                                         -0.28                      -0.34   
4                                         -0.28                      -0.31   

   Total_Monto_Microcredito  Total_Numero_Credito de consumo de bajo monto  \
0                      0.08                                           8.99   
1                     -0.14                                          -0.12   
2                     -0.13                                          -0.12   
3                     -0.14                                          -0.12   
4                     -0.13                                          -0.12   

   Total_Monto_Credito de consumo de bajo monto  \
0                                          8.95   
1                                         -0.12   
2                                         -0.12   
3                                         -0.12   
4                                         -0.12   

   Total_Numero_Credito de vivienda  Total_Monto_Credito de vivienda  \
0                          

# Seleccion de variables

# Construccion Modelo No Supervisado

## Escalado de los datos

In [ ]:
# Seleccion variables
X = df7.drop(['PB_DANE', 'CodigoD', 'Municipio'], axis=1)
y = df7['PB_DANE']

In [ ]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [ ]:
df_X = pd.DataFrame(X_scaled, columns=X.columns)
df_y = pd.DataFrame(y, columns=["PB_DANE"])

In [ ]:
df = df_y.join(df_X)
df.head()

## Desarrollo modelo

### K-Means

In [ ]:
kmeans_1 = KMeans(n_clusters = 5, random_state = 42, n_init=10).fit(df_X)

In [ ]:
# Obtener las etiquetas de cluster generadas por KMeans
cluster_labels_1= kmeans_1.labels_

In [ ]:
crear_grafico_dispersion(kmeans_1, df_X, df_y)

In [ ]:
print("Metrica de silhouette_score: ",silhouette_score(df_X, kmeans_1.labels_))
print("Metrica de Calinski-Harabasz Index: ", calinski_harabasz_score(df_X, kmeans_1.labels_))
print("Metrica de Inercia de k-means: ",kmeans_1.inertia_)

### K-Medoides

In [ ]:
Kmedoides_1 = KMedoids(n_clusters = 5, random_state = 42).fit(df_X)

In [ ]:
cluster_labels_2= Kmedoides_1.labels_

In [ ]:
crear_grafico_dispersion(Kmedoides_1, df_X, df_y)

In [ ]:
print("Metrica de silhouette_score: ",silhouette_score(df_X, Kmedoides_1.labels_))
print("Metrica de Calinski-Harabasz Index: ", calinski_harabasz_score(df_X, Kmedoides_1.labels_))
print("Metrica de Inercia de k-means: ",Kmedoides_1.inertia_)

## Calibracion parámetros

### K-Means

In [ ]:
# Numero de cluster a validar
rg=50

# Inicializar estructuras necesarias para la seleccion de cluster optimo
varianza_intra_cluster = []
silhouettes = []

# iterar entre todos los posibles clusters seleccionados para encontrar el optimo
for k in range (1, rg): # Evaluamos entre 1 a rg clusters posibles
    kmeans = KMeans(n_clusters = k, random_state = 42, n_init=50).fit(df_X)
    varianza_intra_cluster.append(kmeans.inertia_)
    try:
        silhouette = silhouette_score(df_X, kmeans.labels_)
    except:
        silhouette = 0 # El índice de Silhouette sólo se puede calcular cuando se tiene K>1
    silhouettes.append(silhouette)

In [ ]:
# Graficar la varianza intra cluster por cada uno de las posibles combinaciones de cluster a validar
plt.plot(range(1, rg), varianza_intra_cluster, marker='o')
plt.xlabel('Número de clústeres (K)')
plt.ylabel('Varianza intra clúster')
plt.show()

In [ ]:
# Graficar el indice de sihoutette por cada uno de las posibles combinaciones de cluster a validar
plt.plot(range(1, rg), silhouettes, marker='o')
plt.xlabel('Número de clústeres (K)')
plt.ylabel('Índice de Silhouette')
plt.grid()
plt.show()

### K-Medoides

In [ ]:
# Numero de cluster a validar
rg=50

# Inicializar estructuras necesarias para la seleccion de cluster optimo
varianza_intra_cluster = []
silhouettes = []

# iterar entre todos los posibles clusters seleccionados para encontrar el optimo
for k in range (1, rg): # Evaluamos entre 1 a rg clusters posibles
    kmedoids = KMedoids(n_clusters = k, random_state = 42).fit(df_X)
    varianza_intra_cluster.append(kmedoids.inertia_)
    try:
        silhouette = silhouette_score(df_X, kmedoids.labels_)
    except:
        silhouette = 0 # El índice de Silhouette sólo se puede calcular cuando se tiene K>1
    silhouettes.append(silhouette)

In [ ]:
# Graficar la varianza intra cluster por cada uno de las posibles combinaciones de cluster a validar
plt.plot(range(1, rg), varianza_intra_cluster, marker='o')
plt.xlabel('Número de clústeres (K)')
plt.ylabel('Varianza intra clúster')
plt.show()

In [ ]:
# Graficar el indice de sihoutette por cada uno de las posibles combinaciones de cluster a validar
plt.plot(range(1, rg), silhouettes, marker='o')
plt.xlabel('Número de clústeres (K)')
plt.ylabel('Índice de Silhouette')
plt.grid()
plt.show()

## Mejor modelo

In [ ]:
kmeans_2 = KMeans(n_clusters = 6, random_state = 42, n_init=50).fit(df_X)

In [ ]:
crear_grafico_dispersion(kmeans_2, df_X, df_y)

In [ ]:
print("Metrica de silhouette_score: ",silhouette_score(df_X, kmeans_2.labels_))
print("Metrica de Calinski-Harabasz Index: ", calinski_harabasz_score(df_X, kmeans_2.labels_))
print("Metrica de Inercia de k-means: ",kmeans_2.inertia_)

# Analisis Clusters - Mejor Modelo

## Analisis - clusters a data set original

In [ ]:
# Obtener las etiquetas de cluster generadas por KMeans
X_clusters = df7
X_clusters.drop(columns=['PB_DANE', 'CodigoD'], inplace=True)
X_clusters.insert(0, 'clusters', kmeans_2.labels_)

# ver df con labels de clusters
X_clusters.head()

In [ ]:
# Revisar los municipios con mayor indice en que grupo quedaron
X_clusters.sort_values(by='INDICADOR', ascending=False).head(20)

In [ ]:
# describe de los grupos
X_clusters.groupby('clusters').describe()

## Conclusion - Clusters

Se identifican 6 clusters de los cuales se resaltan el cluster 3 con un solo municipio el cual es el distrito capital Bogota el cual no sera tomando en cuenta ya que su desarrollo y la inclusion digital que este presenta tiene varios factores especiales con respecto a los demas.

El cluster 5 conformado por 10 municipios los cuales presentan un indicador promedio de 0.33 y un particular porcentaje de participacion en educacion muy bajo siendo de 0.04 en promedio.

y por ultimo el cluster 1 con 22 municipios el cual presenta un indicador promedio de 0.28 y un porcentaje de participacion de educacion mucho mejor el cual es de 0.59 en promedio.

con estos dos clusters podemos evidenciar las 2 situaciones contrarias en el pais donde el indicador ya se encuentra posicionado y con los correctos planes o iniciativas en cada uno podemos mejorar la integracion con las comunidades teniendo en cuenta las necesidades y el contexto donde se encuentran los 2 clusters.